### PR TITLE
Adding Tests

### PR DESCRIPTION
**Pull request by HimanshuJanbandhu**
I've added tests for autoregressive_generator.py, llama_model_utils.py, self_speculation_generator.py, speculative_streamer.py.

- There are some mistakes in self_speculation_generator. Since I've a low spec system wasn't able to test properly.
- Found a mistake in delete function in speculative_streamer. Plaese take a look.
- All reamaing tests are passed.

### CODE REVIEW

File: self_speculation/__pycache__/__init__.cpython-310.pyc
Comment (mostafaelhoushi) [inline self_speculation/__pycache__/__init__.cpython-310.pyc:1]: Is it possible to remove the *.pyc files from the PR?
I think you can run: 
```
git rm -r self_speculation/__pycache__
git rm -r tests/__pycache__
```

File: requirements.txt [hunk 1]
```diff
@@ -11,3 +11,5 @@ tabulate==0.9.0
 colorama==0.4.6
 protobuf==5.28.2
 lm_eval==0.4.5
+pytest==8.3.3
+pytest-mock==3.14.0
```
Comment (mostafaelhoushi) [inline requirements.txt:15]: Can you add an empty new line here?

File: tests/test_autoregressive_generator.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
+import pytest
```
Comment (mostafaelhoushi) [inline tests/test_autoregressive_generator.py:1]: Can you add the Meta Copyright header to any file added to the PR?
You can copy and paste it from other files:

```
# Copyright (c) Meta Platforms, Inc. and affiliates.
# All rights reserved.
#
# This source code is licensed under the license found in the
# LICENSE file in the root directory of this source tree.
#

```

**Review by mostafaelhoushi — COMMENTED** [review]
Thanks @HimanshuJanbandhu for working on this PR! I highly appreciate that.
I just made a quick look and made some initial feedback.

Still need to look deeper and check the bug that you referred to in the streamer.

**Review by HimanshuJanbandhu — COMMENTED** [review]
Made initial changes

**Comment by HimanshuJanbandhu** [comment]
> Still need to look deeper and check the bug that you referred to in the streamer.
My test for the delete function is failing, the text_cache is not getting cleared by delete function.

**Comment by HimanshuJanbandhu** [comment]
Also if you want to run any tests, use this

```
$ pytest tests/<test_file>
```

File: __pycache__/arguments.cpython-310.pyc
Comment (mostafaelhoushi) [inline __pycache__/arguments.cpython-310.pyc:1]: Please also remove this .pyc file

File: requirements.txt [hunk 3]
```diff
@@ -11,3 +11,7 @@ tabulate==0.9.0
...
+\n
+
```
Comment (mostafaelhoushi) [inline requirements.txt:17]: Please remove `\n`
```suggestion
```

File: tests/test_autoregressive_generator.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+@pytest.fixture
+def model_and_config():
+    """Fixture to create a model and a generation configuration."""
+    # Initialize a lightweight version of LlamaForCausalLM if possible
+    tokenizer = AutoTokenizer.from_pretrained(local_model_path)
+    model = AutoModelForCausalLM.from_pretrained(
+        local_model_path,
+        use_safetensors=True,
+        device_map="mps",
```
Comment (mostafaelhoushi) [inline tests/test_autoregressive_generator.py:27]: Can we replace the `mps` with `auto`? 
I was running on a Linux server with a GPU, and `auto` automatically detects the GPU. I think when running on MacBook, it should automatically detect `mps`, but you can give it a try.

```suggestion
        device_map="auto",
```

File: tests/test_llama_model_utils.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    logits = torch.tensor([[[0.1, 0.2, 0.7]]])
+    next_token, _ = decode_next_token(logits)
+    assert next_token.item() == 2
+
+def test_decode_next_token_sampling():
+    """Test sampling decoding"""
+    logits = torch.tensor([[[0.1, 0.2, 0.7]]])
+    next_token, probabilities = decode_next_token(logits, sample=True)
+    assert next_token.shape == torch.Size([1, 1])
+    assert probabilities.shape == torch.Size([1, 3])
```
Comment (mostafaelhoushi) [inline tests/test_llama_model_utils.py:76]: Please add empty new line at the end of the file

File: tests/test_self_speculation_generator.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+@pytest.fixture
+def model_and_config():
+    """Fixture to create a model and a generation configuration."""
+    # Initialize a lightweight version of LlamaForCausalLM if possible
+    tokenizer = AutoTokenizer.from_pretrained(local_model_path)
+    model = AutoModelForCausalLM.from_pretrained(
+        local_model_path,
+        use_safetensors=True,
+        device_map="mps",
```
Comment (mostafaelhoushi) [inline tests/test_self_speculation_generator.py:28]:
```suggestion
        device_map="auto",
```

File: tests/tests_constants.py [hunk 7]
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the license found in the
+# LICENSE file in the root directory of this source tree.
+#
+
+
+local_model_path = "facebook/layerskip-llama3.2-1B"
```
Comment (mostafaelhoushi) [inline tests/tests_constants.py:9]: Please add empty new line at the end of the file

**Review by mostafaelhoushi — COMMENTED** [review]
Thanks again and sorry for the delay from my side.
I just tried to run it locally. I made some changes to get it to work that I included in this round of feedback, but I am getting these errors:
```
========================================================================= short test summary info ==========================================================================
FAILED tests/test_self_speculation_generator.py::test_single_step_speculation_handling_eos - RuntimeError: The expanded size of the tensor (3) must match the existing size (2) at non-singleton dimension 3.  Target sizes: [1, 32, 2, 3].  Tensor sizes: [1, 1, 2, 2]
FAILED tests/test_self_speculation_generator.py::test_generate_token_ids_with_logit_processors - ZeroDivisionError: division by zero
FAILED tests/test_speculative_streamer.py::test_delete_tokens - AssertionError: assert '1 2 3 4 5' == '1 2 3'
====================================================================== 3 failed, 13 passed in 23.87s =======================================================================
```

You can see the full log [here](https://gist.github.com/mostafaelhoushi/ea5c0c8cb1e37755f3e09223ac583c31). Are you getting similar errors as well?

**Comment by mostafaelhoushi** [comment]
Sorry, I forgot that you mentioned you are also getting an error with `test_delete_tokens`. I want to double check, are you getting errors with `test_single_step_speculation_handling_eos` and `test_generate_token_ids_with_logit_processors`?

File: tests/test_speculative_streamer.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    streamer.put(input_tensor, is_draft=False)
+    streamer.tokenizer.decode.assert_called_once_with([1, 2, 3], **streamer.decode_kwargs)
+    assert streamer.text_cache == '1 2 3'
+
+def test_delete_tokens(streamer):
+    input_tensor = torch.tensor([1, 2, 3, 4, 5])
+    streamer.put(input_tensor, is_draft=False)
+    
+    streamer.delete(2)
+    assert streamer.text_cache == '1 2 3'
```
Comment (mostafaelhoushi) [inline tests/test_speculative_streamer.py:45]: I think what is happening here is that `delete_tokens` doesn't really delete tokens from the end of the cache, but rather reduces the length of the cache, and hence any tokens at `index > streamer.print_len` are not really counted for (as if they are not there) and the streamer will not print them.

So when checking the contents of the streamer we need to check the cache up to `streamer.print_len`.
```suggestion
    assert streamer.text_cache[:streamer.print_len] == '1 2 3'
```

**Comment by HimanshuJanbandhu** [comment]
> Sorry, I forgot that you mentioned you are also getting an error with `test_delete_tokens`. I want to double check, are you getting errors with `test_single_step_speculation_handling_eos` and `test_generate_token_ids_with_logit_processors`?

Yes I'm getting errors with `test_single_step_speculation_handling_eos` and `test_generate_token_ids_with_logit_processors`, both of these functions are in self_speculation_generator.py
I'm not able to understand if I'm testing it wrong or the function has a mistake.

This error is happening because no token is getting generated, that's why there's a `ZeroDivisionError`. We should make a test that generates some tokens, but at the same time this `ZeroDivisionError` shouldn't occur.
```
FAILED tests/test_self_speculation_generator.py::test_generate_token_ids_with_logit_processors - ZeroDivisionError: division by zero
```

I'm working on new commit and `test_delete_tokens` I've resolved

**Review by HimanshuJanbandhu — COMMENTED** [review]
Corrected `test_delete_tokens` in `test_speculative_streamer.py`.
Made other minor changes, tests in `test_self_speculation_generator.py` still failing

File: tests/test_autoregressive_generator.py [hunk 9]
```diff
@@ -60,3 +60,5 @@ def test_generate_token_ids_with_logit_processors(model_and_config):
     assert len(result.predicted_tokens) > 0
     assert eos_token_id in result.predicted_tokens or len(result.predicted_tokens) == config.max_steps
 
+
+
```
Comment (mostafaelhoushi) [inline tests/test_autoregressive_generator.py:64]: Let's aim to have 1 empty new line at the end of each file
```suggestion
```

File: tests/test_llama_model_utils.py [hunk 10]
```diff
@@ -73,4 +73,7 @@ def test_decode_next_token_sampling():
     logits = torch.tensor([[[0.1, 0.2, 0.7]]])
     next_token, probabilities = decode_next_token(logits, sample=True)
     assert next_token.shape == torch.Size([1, 1])
-    assert probabilities.shape == torch.Size([1, 3])
\ No newline at end of file
+    assert probabilities.shape == torch.Size([1, 3])
+
+
+
```
Comment (mostafaelhoushi) [inline tests/test_llama_model_utils.py:79]:
```suggestion
    assert probabilities.shape == torch.Size([1, 3])

```

File: tests/test_self_speculation_generator.py [hunk 11]
```diff
@@ -79,3 +79,5 @@ def test_generate_token_ids_with_logit_processors(model_and_config):
     assert eos_token_id in result.predicted_tokens or len(result.predicted_tokens) == config.max_steps
 
 
+
+
```
Comment (mostafaelhoushi) [inline tests/test_self_speculation_generator.py:83]:
```suggestion

```

File: tests/test_speculative_streamer.py [hunk 12]
```diff
@@ -52,3 +52,6 @@ def test_draft_handling(streamer, is_draft):
         assert '\n' not in streamer.text_cache
     else:
         assert streamer.text_cache == '1 2 3'
+
+
+
```
Comment (mostafaelhoushi) [inline tests/test_speculative_streamer.py:57]:
```suggestion

```

**Review by mostafaelhoushi — COMMENTED** [review]
Added some comments to try to ensure we have one empty new line at the end of each file

File: tests/test_self_speculation_generator.py [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+    tokenizer = AutoTokenizer.from_pretrained(local_model_path)
+    model = AutoModelForCausalLM.from_pretrained(
+        local_model_path,
+        use_safetensors=True,
+        device_map="auto",
+        torch_dtype=torch.float,
+    )
+
+    model.eval()
+    config = GenerationConfig(max_steps=4)
```
Comment (mostafaelhoushi) [inline tests/test_self_speculation_generator.py:33]: I did this change locally and it seems to fix the 2 errors.
I think the reason is taht by default `num_speculations` is -1 and hence there are no draft generations and that caused the numerical errors.
```suggestion
    config = GenerationConfig(max_steps=4, exit_layer=4, num_speculations=4)
```

**Review by HimanshuJanbandhu — COMMENTED** [review]
Removed extra lines and made changes in test_self_speculation_generator.py, now my tests are running fine. 
Thanks for your help!

**Review by mostafaelhoushi — COMMENTED** [review]
Thanks @HimanshuJanbandhu ! I made one last round of review and noticed there's a .pyc file that got checked in by mistake.

When it's removed I can merge the PR

**Review by HimanshuJanbandhu — COMMENTED** [review]
Hey, my bad. Missed this pycache by mistake.
This commit looks good.

**Comment by mostafaelhoushi** [comment]
Thanks again @HimanshuJanbandhu for all your help! I highly appreciate that.

!closes #7

**Comment by HimanshuJanbandhu** [comment]
Thanks for the opportunity @mostafaelhoushi!

**Pull request merged**